### PR TITLE
Implement DPSGD example.

### PR DESCRIPTION
**Pull request by n2cholas**
DPSGD example as requested in #53. Verified the accuracies are close with a few seeds [here](https://colab.research.google.com/drive/107SrfBIyZdD9ES67xEvuNy0J8kUEirl_?usp=sharing). This is based on the [`differentially_private_sgd.py`](https://github.com/google/jax/blob/master/examples/differentially_private_sgd.py) example in the JAX repo, but refactored to better match Optax's style.

Questions:
1. How do I run these examples? I couldn't get the `from optax.examples import datasets` import to work when running it via `python3 dpsgd_example.py`
2. Is the name ok? The JAX example named it `differentially_private_sgd.py` I used `dpsgd_example.py` because `differentially_private_sgd_example.py` is kind of long.
3. The default seed of 0 results in a slightly lower accuracy than the reported (just like we saw in #53), should I switch the default seed?
4. Should I create a `requirements.txt` specifically for this example? Or is the comment in the file enough (that specifies how to install `tensorflow_privacy`.

Additional questions/concerns listed as review comments below.

cc: @mtthss

### CODE REVIEW

File: examples/dpsgd_example.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
+# Copyright 2021 DeepMind Technologies Limited. All Rights Reserved.
```
Comment (n2cholas) [inline examples/dpsgd_example.py:1]: Is this copyright header ok? A big portion of the code is from the JAX repo.

File: examples/dpsgd_example.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+This script contains a JAX implementation of Differentially Private Stochastic
+Gradient Descent (https://arxiv.org/abs/1607.00133). DPSGD requires clipping
+the per-example parameter gradients, which is non-trivial to implement
+efficiently for convolutional neural networks.  The JAX XLA compiler shines in
+this setting by optimizing the minibatch-vectorized computation for
+convolutional architectures. Train time takes a few seconds per epoch on a
+commodity GPU.
+This code depends on tensorflow_privacy (https://github.com/tensorflow/privacy)
+  Install instructions:
+    $ pip install tensorflow git+https://github.com/tensorflow/privacy.git
```
Comment (n2cholas) [inline examples/dpsgd_example.py:25]: Note: these installation instructions are slightly different from the [original JAX version](https://github.com/google/jax/blob/master/examples/differentially_private_sgd.py#L27-L30).

File: examples/dpsgd_example.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+   --epochs=60 \
+   --learning_rate=.15 \
+  # this private baseline should get ~97% acc
+  python -m examples.differentially_private_sgd \
+   --dpsgd=True \
+   --noise_multiplier=0.7 \
+   --l2_norm_clip=1.5 \
+   --epochs=45 \
+   --learning_rate=.25 \
+"""
```
Comment (n2cholas) [inline examples/dpsgd_example.py:55]: TODO: Change these after figuring out the correct way to run the code.

File: examples/dpsgd_example.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+from absl import app
+from absl import flags
+
+import jax
+import jax.numpy as jnp
+from jax.experimental import stax
+import optax
+
+import datasets
```
Comment (n2cholas) [inline examples/dpsgd_example.py:68]: TODO: Change this import to `from optax.examples import datasets` after figuring out how to run the code.

File: examples/dpsgd_example.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                   'Ratio of the standard deviation to the clipping norm')
+flags.DEFINE_float('l2_norm_clip', 1.0, 'Clipping norm')
+flags.DEFINE_integer('batch_size', 256, 'Batch size')
+flags.DEFINE_integer('epochs', 60, 'Number of epochs')
+flags.DEFINE_integer('seed', 0, 'Seed for jax PRNG')
+flags.DEFINE_integer(
+    'microbatches', None, 'Number of microbatches '
+    '(must evenly divide batch_size)')
+flags.DEFINE_string('model_dir', None, 'Model directory')
+flags.DEFINE_float('delta', 1e-5, 'Target delta used to compute privacy spent.')
```
Comment (n2cholas) [inline examples/dpsgd_example.py:91]: NOTE: Not `delta` flag not present in original JAX example, added because it was easy to, could remove if we want to stay closer to the original.

File: examples/dpsgd_example.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+@jax.jit
+def test_step(params, batch):
+  loss, logits = loss_fn(params, batch)
+  accuracy = (logits.argmax(1) == batch['label'].argmax(1)).mean()
+  return loss, accuracy * 100
+
+
+def main(_):
+  train_dataset = datasets.load_image_dataset('mnist', FLAGS.batch_size)
+  train_dataset = list(train_dataset.as_numpy_iterator())
```
Comment (n2cholas) [inline examples/dpsgd_example.py:132]: When used in the training loop as `train_dataset.as_numpy_iterator()`, the time per private epoch is ~4 seconds, which is significantly slower than this using list (which results in a per private epoch time of ~0.7 seconds). I tried using `prefetch`, which didn't help. How else can I improve this?

File: examples/dpsgd_example.py (hunk 1 above)
Comment (mtthss) [reply examples/dpsgd_example.py:1]: Yes this is fine, but please add an attribution in the file doc-string explaining that a big portion of the code is forked from the JAX example

File: examples/dpsgd_example.py (hunk 2 above)
Comment (mtthss) [reply examples/dpsgd_example.py:25]: LGTM

**Review by mtthss — COMMENTED** [review]
2. The name is good
3. Feel free to change the seed
4. Any downside in just adding it to `requirements/requirements-examples.txt`? This is not automatically pulled in when installing optax so its fine to depend on more things in the examples

File: examples/dpsgd_example.py (hunk 1 above)
Comment (n2cholas) [reply examples/dpsgd_example.py]: Done!

File: examples/dpsgd_example.py (hunk 2 above)
Comment (n2cholas) [reply examples/dpsgd_example.py]: Will resolve this because I moved the installation to `requirements/requirements-examples.txt`

**Comment by n2cholas** [comment]
> 1. The name is good

I switched the name back to `differentially_private_sgd.py`, since it's in the `examples/` directory, I thought the `_example` suffix was redundant.

> 2. Feel free to change the seed

Done!

> 3. Any downside in just adding it to `requirements/requirements-examples.txt`? This is not automatically pulled in when installing optax so its fine to depend on more things in the examples

That sounds good, done! I also needed to add `tensorflow` and `tensorflow_datasets` to that file to make it complete.

I also reran the colab with the exact commands specified in the docstring, with the dependencies installed via `requirements/requirements-examples.txt` instead of manually.

**Review by mtthss — APPROVED** [review]

**Comment by mtthss** [comment]
Apologies for the delay in merging.
There is an issue with the ordering of imports
Could you update it to:
```
import time
import warnings

from absl import app
from absl import flags
import datasets
import jax
from jax.experimental import stax
import jax.numpy as jnp
import optax
from tensorflow_privacy.privacy.analysis.rdp_accountant import compute_rdp
from tensorflow_privacy.privacy.analysis.rdp_accountant import get_privacy_spent
```

**Comment by n2cholas** [comment]
> import time
> import warnings
> 
> from absl import app
> from absl import flags
> import datasets
> import jax
> from jax.experimental import stax
> import jax.numpy as jnp
> import optax
> from tensorflow_privacy.privacy.analysis.rdp_accountant import compute_rdp
> from tensorflow_privacy.privacy.analysis.rdp_accountant import get_privacy_spent

Sure, done! I want to confirm that `import datasets` is ok, since the other examples use `from optax.examples import datasets`. The only reason I used `import datasets` instead is because I couldn't figure how to run the examples with the `from optax.examples import datasets` import.

**Comment by 8bitmp3** [comment]
Looks good @n2cholas. Noticed the PR is a bit stale, @mtthss - does it need to be reapproved since 15 April? Thanks all, I'm learning a lot

**Review by mtthss — APPROVED** [review]
